### ISSUE TITLE
Don't trigger translator detection on iframes

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#1021]
It seems like we're doing a huge amount of extra work on advertising iframes. Are there legit cases where we actually need to check iframes? Can we just not? Alternatively, can we allow translators to specify when they need to be run on iframes, only when the target regexp matches (so that we'd do regexp tests on all iframes but not run `detectWeb()` for all generic translators)?

/cc @simonster @zotero/translator-reviewers

**Comment by zuphilip** [comment]
There are legit cases where we have to check iframes during translation, e.g. [LexisNexis](https://github.com/zotero/translators/blob/master/LexisNexis.js). Some more can be found with [this search in the repo](https://github.com/zotero/translators/search?utf8=%E2%9C%93&q=iframe). However, I am not sure, that we need the iframes for detection already or just for during the scraping part of the translators.

**Comment by dstillman** [comment]
OK, yeah, so Lexis is definitely a case where we need to run detection on the iframe. But that's an example where it seems like the translator could indicate that it needs the regexp tested against all iframes, which it could even do with a `rootTarget` property that matched the root URL. We never really worried about regexp performance, but we now have almost 500 translators, and if there can now be dozens of iframes per page, that's a lot of unnecessary tests that might actually affect performance.

But the low-hanging fruit here would be to avoid running detection for the generic translators when in a non-root frame. That would mean we wouldn't detect DOI/COinS/etc. in frames, but that seems like a fair trade-off for avoiding a lot of work on ad-heavy pages.

I think we could do this by setting a flag on Zotero.Translate.Web in browser.js::detectTranslators() when `doc != rootDoc`, and Zotero.Translate.Web.prototype._getTranslatorsGetPotentialTranslators() would pass a `requireTarget` argument to Zotero.Translators.getWebTranslatorsForLocation(), causing it to skip the generic translators.

This isn't totally representative of the gains we'd see, but adding

```javascript
if (rootDoc != doc) {
    Zotero.debug("Ignoring non-root doc");
    return;
}
```

after the first block in `browser.js::contentLoad()` avoids 300ms in translate_firefox.js on some sites on my system (I think from wrapping the document object and such — not sure how performance in the connector would compare).

**Comment by dstillman** [comment]
OK, so I'd like to move forward with this, by having translators that might need to run in iframes declare it in their metadata with a `rootTarget` property. If `rootTarget` exists and matches the root doc's URL, `target` would run against all iframes of that doc, and any that matched would run `detectWeb`. If `rootTarget` doesn't exist or is empty, iframes would be ignored. Notably, generic translators would not have `rootTarget` and therefore wouldn't run `detectWeb` in iframes.

It looks like a bunch of translators access iframes from the root doc, but those wouldn't be affected — this would only affect translators that actually need to run `detectWeb` in the iframe and can't be rewritten to run from the root doc.

@simonster, @adam3smith, any concerns here? Do we know off-hand which translators other than Lexis actually need to run in the iframe?

**Comment by dstillman** [comment]
Slightly revised plan: `target` and `targetAll`

**`target` with no or empty `targetAll`**: check root doc against `target`; if it matches, run `detectWeb()`; iframes are ignored; this will be the default for all unmodified existing translators

**`target` with non-empty `targetAll`**: check root doc against `target`; if it matches, check all documents on page — root and iframes — against `targetAll`; if any match, run `detectWeb()` on them; a root doc that matches `target` only will not run `detectWeb()`

Given that root docs can generally access iframes from the same domain, `targetAll` should probably usually be for a different domain. (If it was also necessary to match other, non-iframe pages on the original domain, `targetAll` could be made to cover both domains.)

The reason for two regexps, instead of just a boolean to enable iframe-checking for root docs matching `target`, is to avoid making unnecessary `detectWeb()` calls. With a boolean, `target` might need to be broad to match the root doc, but then `detectWeb()` would possibly need to run on a large number of iframes on the site that match the same pattern. With two separate targets, there can be a broad regexp to identify pages that need iframes checked, and then those iframe URLs can be compared against a much more specific regexp, ideally only running `detectWeb()` on a specific frame.

**Comment by simonster** [comment]
No concerns from me. I think frames/iframes are pretty uncommon, although I vaguely remember a library catalog that used frames.

**Comment by adam3smith** [comment]
Plan sounds good. I'm sure there are a couple of old translators that detect in iframes (I haven't come across any in the last 2 years or so, but I do remember doing this occasionally "in the old days"), so we'll just have to lookout for error reports.

**Comment by aurimasv** [comment]
IIRC Highwire 2 detects a particular iframe within a page (for PDF pages),
but I believe it is an empty target, so sounds like it should not be
affected.

**Comment by dstillman** [comment]
We should check those, but the translators that actually mention "iframe" might actually be less likely to be affected here (unless it's in a comment), since there's a good chance those are triggering on the parent doc and reaching into the iframe, which won't be affected. The ones we need to find are the ones that are running in the iframe.

Hopefully we'll have this enabled on the 5.0 beta soon and we can start adding the `targetAll` property.

**Comment by dstillman** [comment]
`targetAll` support is now implemented and available in r49 of the 5.0 beta. This means that any translators that rely on detection in iframes will no longer work in 5.0 unless a `targetAll` property is added and `target` is changed to match the root doc. While 4.0 is still around, `target` might need to be broader on those sites (or use a `|`), since it needs to match both the root doc (for 5.0) and the iframe (for 4.0). Ultimately, it'd be better to rewrite those to just access the iframes from the root doc where possible.

I'll also switch the translator tests to run on 5.0 so we can start seeing where this is broken.
